### PR TITLE
Real-time LF sampling

### PR DESCRIPTION
**Pull request by wh201906**
1. The client side doesn't handle these samples. To get the samples, you need to run socat with -x option to log the data, then plot it somewhere.
2. ~~When starting bulk tranfer, \~3 samples will lost due to the time cost.~~ Some of the USB transfer payload can be distributed into every sample. (Done in 523fde21a4ce0e62ed299cce3ed2cb6814f67ffa)
3. ~~I use the verbose option as a flag for real-time sampling~~ (`-r` for real-time sampling)

### CODE REVIEW

**Comment by wh201906** [comment]
> 2\. When starting bulk tranfer, ~3 samples will lost due to the time cost

Maybe this is caused by the `-x` option in `socat`
Test `hw status`:
```
socat -dd /dev/ttyACM0,raw,echo=0,b115200 udp-listen:18888,fork,reuseaddr
Transfer Speed PM3 -> Client... 732160 bytes/s

socat -dd -x /dev/ttyACM0,raw,echo=0,b115200 udp-listen:18888,fork,reuseaddr
Transfer Speed PM3 -> Client... 325354 bytes/s

stdbuf -oL socat -dd -x /dev/ttyACM0,raw,echo=0,b115200 udp-listen:18888,fork,reuseaddr &> lfread3.txt
Transfer Speed PM3 -> Client... 304956 bytes/s

# Can be optimized by this
stdbuf -oL socat -dd -x /dev/ttyACM0,raw,echo=0,b115200 udp-listen:18888,fork,reuseaddr &> /mnt/ramdisk/lfread3.txt
Transfer Speed PM3 -> Client... 675840 bytes/s
```

**Comment by iceman1001** [comment]
Interesting
However,  all those USB stuff doesn't belong in the lfsampling...
you need to refactor it

**Comment by wh201906** [comment]
Yes I will. The commits there is only for test now. I'll clean up the code and squash the commits once everything is done.

**Comment by iceman1001** [comment]
I am happy to see your great efforts of making the proxmark3 run better with new functions.

File: armsrc/lfsampling.c
Comment (henrygab) [inline armsrc/lfsampling.c:523]: After button is pressed, or the loop exist for another reason, could there be one or more samples that were pushed into the hardware FIFO buffer?  If so, should a final call to `usb_write_request()` occur?
Careful ... also have to consider error exit paths ... which may indicate those final bytes are better off lost?

File: common_arm/usb_cdc.h
Comment (henrygab) [inline common_arm/usb_cdc.h:33]: First, I admit I may be fully misunderstanding something.

<details><summary>These functions appear to have implicit hidden relationship</summary><p/>

It seems that these two new functions have an implicit relationship, which is not obvious from their names.
Specifically, it would appear that the following sequence is necessary:

1. At function entry, clearing any prior transfers / data in USB FIFO buffer.  No function for this currently exists.
2. Calling `usb_write_byte_async()` ... which really just adds a byte to the hardware FIFO for the next transfer?
3. Tracking how many bytes are in the hardware FIFO buffer (e.g., how many times called above function....)
4. Before hardware FIFO fills, call `usb_write_request()` ... which may spin / cause lost samples ....
5. At function exit, if hardware FIFO is not empty, send final (partial) packet of data to empty FIFO buffer.
    (no function exists for this at the moment?)

</details> 

<details><summary>Naive code added later based on current names</summary><p/>

To show how the names may suggest functionality, such as presuming that async writes would eventually send properly, but result in really unexpected and difficult-to-parse problems...

```
// send buffer contents twice
for (int i = 0; i < 512; i++) {
    usb_write_byte_async(buffer[i]); // definitely does not result in sending 512 bytes
}
usb_write_request(); // b/c hardware FIFO is < 512 bytes (328 bytes ... likely split across all endpoints?)
for (int i = 0; i < 512; i++) {
    // maybe user doesn't even recognize need to call usb_write_request()...
    usb_write_byte_async(buffer[i]); 
}
usb_write(buffer, 512); // who knows what odd results will happen here?
usb_write_byte_async(0x77); // this byte might be lost, or corrupt the next packet?
usb_write(buffer, 512);
usb_write_request(); // what would this do?
```
</details> 

### Questions for discussion....

1. How can this be structured to make clear the hidden bindings between these functions?
    1. For example, that what is currently named `usb_write_byte_async()` requires a call to what is currently named `usb_write_request()`, before other calls that write to USB?
1. Could the answer be as simple as renaming the functions and providing inline function documentation?
1. What happens if too many bytes are attempted to be written asynchronously?
    1. Does the CPU hang?
    1. Is the last byte discarded?
    1. Is the first byte in the FIFO discarded?
    1. Other?

**Review by henrygab — CHANGES_REQUESTED** [review]
First, great job!  This looks like a clean method of using the hardware FIFO with a single collection buffer.

I do have a number of recommended changes.  Some are defensive coding, which are recommended because this is a larger, multi-author project.   Some are minor formatting nitpicks.   

The two new usb function names ... that's a bigger discussion.   I stumbled over what those functions were doing initially, and even now am not 100% sure I get it.

Finally, I could use help understanding why the decimation setting should alter the maximum bytes per USB transfer (details in discord).

Overall, I can't wait for this feature to hit release!

File: armsrc/lfsampling.c
Comment (wh201906) [reply armsrc/lfsampling.c:503]: Got it

File: armsrc/appmain.c
Comment (wh201906) [reply armsrc/appmain.c:864]: Got it

File: armsrc/appmain.c
Comment (wh201906) [reply armsrc/appmain.c:897]: Got it

File: common_arm/usb_cdc.c
Comment (wh201906) [reply common_arm/usb_cdc.c:837]: Got it

File: armsrc/appmain.c
Comment (wh201906) [reply armsrc/appmain.c:889]: May I use `uint8_t` there?

File: client/src/cmdlf.h
Comment (wh201906) [reply client/src/cmdlf.h:44]: For `lf_read_internal()`?

File: armsrc/lfsampling.c
Comment (wh201906) [reply armsrc/lfsampling.c:523]: I haven't implement this part yet. I need to design the overall behavior when the user-required exit or error occurs for both the client and the firmware, so I might do it when the basic sampling code for client and the firmware is finished.

File: client/src/cmdlf.c
Comment (wh201906) [reply client/src/cmdlf.c:707]: I see. I'll do that in a dedicated commit when cleaning up the commit history.

File: common_arm/usb_cdc.h
Comment (wh201906) [reply common_arm/usb_cdc.h:33]: > which really just adds a byte to the hardware FIFO for the next transfer?
Yes.

I admit that the name of these functions are somehow confusing. These inline functions are just hard-coded before 045a10cb03d060d2448d0e2071e2514adcfb2226, but as @iceman1001 suggested I extract them to a different .c file. I didn't select the name carefully. Would you like to help me to pick proper names for them?

Comment (wh201906) [reply common_arm/usb_cdc.h:33]: > What happens if too many bytes are attempted to be written asynchronously? 

I haven't tested that.

File: armsrc/lfsampling.c
Comment (wh201906) [reply armsrc/lfsampling.c]: It should be `const int8_t size_threshold = size_threshold_table[bits_per_sample];` there.

Comment (wh201906) [reply armsrc/lfsampling.c]: As for the validation, the logSample() inside this function validates `the bits_per_sample`, which is not appropriate because it's time-critical. I'm considering moving it before the sampling starts.

File: armsrc/lfsampling.c
Comment (wh201906) [reply armsrc/lfsampling.c:453]: Got it

File: armsrc/lfsampling.c
Comment (wh201906) [reply armsrc/lfsampling.c:509]: Got it

File: client/src/cmdlf.c
Comment (wh201906) [reply client/src/cmdlf.c:703]: Got it

File: client/src/cmdlf.c
Comment (wh201906) [reply client/src/cmdlf.c:714]: I reuse the old logic there. I think it's better to finish this in the commit of defining packet format in a common header file

File: client/src/cmdlf.h
Comment (iceman1001) [reply client/src/cmdlf.h]: For static functions, we don't do any declarations.

File: client/src/cmdlf.c
Comment (iceman1001) [reply client/src/cmdlf.c:703]: we have used common typedef:d stucts for  cmdlf.c in the header file:  ` pm3_cmd.h`  
Use it for such purpose.

File: common_arm/usb_cdc.h
Comment (iceman1001) [reply common_arm/usb_cdc.h:33]: In the folder  `common_arm\`  you find the usb_cdc.c  / .h  files.   Where these functions belongs.

File: armsrc/lfsampling.c
Comment (wh201906) [reply armsrc/lfsampling.c:467]: Got it

File: armsrc/lfsampling.c
Comment (wh201906) [reply armsrc/lfsampling.c:486]: @iceman1001 I found that you added this comment in https://github.com/RfidResearchGroup/proxmark3/commit/a5833fe39069d2ec50e90ee3637b27ccd76f7d79 . Is this only for RDV4?

File: armsrc/lfsampling.c
Comment (wh201906) [reply armsrc/lfsampling.c:489]: I don't quite understand the comment there. @iceman1001 Could you please give me some advice for updating it?

File: armsrc/lfsampling.c
Comment (iceman1001) [reply armsrc/lfsampling.c:486]: I think we have a test point list in the documents folder some where in a note.
Regardless,   just add a clarifying text:     ` (RDV4) Test point 8... `

File: armsrc/lfsampling.c
Comment (wh201906) [reply armsrc/lfsampling.c:490]: Got it

File: armsrc/lfsampling.c
Comment (wh201906) [reply armsrc/lfsampling.c:496]: Got it

File: armsrc/lfsampling.c
Comment (iceman1001) [reply armsrc/lfsampling.c:467]: and we rarely use specialized  compiler directives,   since we support several OS/compiler versions.   

Cluttering code is _not_  not what we want.

File: armsrc/lfsampling.c
Comment (wh201906) [reply armsrc/lfsampling.c:498]: Got it

File: armsrc/appmain.c
Comment (iceman1001) [reply armsrc/appmain.c:890]: we use the bool in structs ( or rather @doegox  introduced it )  since it was supported on the oldest compiler version we support.
it gives use clarify of code  which we like.

File: armsrc/lfsampling.c
Comment (iceman1001) [reply armsrc/lfsampling.c:490]: Interesting idea,  but No thanks.    We really don't want to clutter down code more than it already is.

File: client/src/cmdlf.c
Comment (iceman1001) [reply client/src/cmdlf.c:710]: Since we have 64 KB ram on device side,   we will never reach 30bits of samples.   _just saying_

Comment (wh201906) [reply client/src/cmdlf.c:710]: Got it

File: client/src/cmdlf.c
Comment (wh201906) [reply client/src/cmdlf.c:778]: Got it

File: armsrc/appmain.c
Comment (wh201906) [reply armsrc/appmain.c:890]: If so I'll keep using `bool`

File: armsrc/appmain.c
Comment (wh201906) [reply armsrc/appmain.c:856]: I think I'll keep `bool` as suggested by @iceman1001 
https://github.com/RfidResearchGroup/proxmark3/pull/2167#discussion_r1390229095

File: armsrc/lfsampling.c
Comment (wh201906) [reply armsrc/lfsampling.c:467]: > and we rarely use specialized compiler directives, since we support several OS/compiler versions.
> 
> Cluttering code is _not_ not what we want.

I think the GCC extensions there is acceptable because we only use `gcc-arm-none-eabi` to compile the code. And the code won't be too messy
from `if ((checked >= 4000) && trigger_hit == false) {`
to `if (unlikely(trigger_hit == false && (checked >= 4000))) {`
@iceman1001 How do you think? The `unlikely()` does balance the length of each branch

File: armsrc/lfsampling.c
Comment (wh201906) [reply armsrc/lfsampling.c:486]: Got it

File: client/src/cmdlf.h
Comment (wh201906) [reply client/src/cmdlf.h]: Got it

File: client/src/cmdlf.c
Comment (wh201906) [reply client/src/cmdlf.c:703]: Got it

**Comment by wh201906** [comment]
In cast that I miss the review advices you provided, I'll apply all of them first and rewrite the commit history as need.

**Comment by wh201906** [comment]
@iceman1001 Do you consider using the `unlikely()` GCC extension? The arm code is always compiled with `gcc-arm-none-eabi` on all platforms. If you don't like it I will revert the changes about them.

**Comment by wh201906** [comment]
Well it looks like the `unlikely()` is undefined. I just tested it with Compile Explorer and it shows it's undefined, just like the CI output.
However, the `__builtin_expect()` does works, so we need to add an extra `#define` for it if we need `unlikely()`

**Comment by iceman1001** [comment]
I seriously doubt that the ___builtin_expect()_  call will improve the loop much. 

and since we compile it on decently recent versions on the gcc-arm-none-eabi compilers (2018-2022 ish) across the OS platforms,   we can add it.     

I would love to see if there is actually some gains,  otherwise these hints adds more confusing when trying to read the code.

**Comment by wh201906** [comment]
I see

File: armsrc/lfsampling.c
Comment (wh201906) [reply armsrc/lfsampling.c:537]: I've finished that. Please check commit 079d2c7e37e116e82010715185e56c13547feb72

File: armsrc/util.c
Comment (iceman1001) [inline armsrc/util.c:315]: you can simplify this to have value only.    if its a unsigned int,  either its zero or a value.
no need to do extra compare.

File: common_arm/usb_cdc.c
Comment (iceman1001) [inline common_arm/usb_cdc.c:657]: I suspect this call makes things slower...

Comment (wh201906) [reply common_arm/usb_cdc.c]: Yeah that's true. The gcc tells me the inlining failed there.

File: armsrc/util.c
Comment (wh201906) [reply armsrc/util.c:315]: Got it

**Comment by wh201906** [comment]
@iceman1001 The samples can be fetched to the client now.
https://github.com/RfidResearchGroup/proxmark3/blob/e44161bb602fe9b5c37daf052eff281f7732521f/client/src/cmdlf.c#L716-L718
The real-time sampling process can be stopped by pressing Enter on the client or pressing the button on PM3. It can also be stopped if it reaches the number of `-s` option.
How can I store them into a file compatible with the `.pm3` format? Is there any existing function for this?

**Comment by wh201906** [comment]
@iceman1001 The real-time `lf read` is working now. However, I added some of the code from `cmddata.c` to `cmdlf.c` to make it work. Plus, the graph buffer is not resizable, so it's hard to plot and save all of the samples. I think we need to refactor the code.

**Comment by iceman1001** [comment]
Nice!

`data save -f`   is the function all.   it save the length of what has been loaded onto the graphbuffer.

**Comment by wh201906** [comment]
Here are some bugs when `bit_per_sample` is not 8
1. `async_usb_write_requestWrite()` always returns false
2. the client crashes after receiving the data

File: armsrc/lfsampling.c
Comment (henrygab) [reply armsrc/lfsampling.c]: It doesn't matter if it's validated later.  If the value exceed 8, this code will have undefined behavior due to out-of-bounds array access.  Therefore, the validation would need to occur prior to that use as an array index.

File: common_arm/usb_cdc.h
Comment (henrygab) [reply common_arm/usb_cdc.h]: > I admit that the name of these functions are somehow confusing. 
> I didn't select the name carefully. Would you like to help me to pick proper names for them?

I think you created much clearer names with commit https://github.com/RfidResearchGroup/proxmark3/commit/079d2c7e37e116e82010715185e56c13547feb72.  👍 

(https://github.com/RfidResearchGroup/proxmark3/blob/079d2c7e37e116e82010715185e56c13547feb72/common_arm/usb_cdc.h#L34-L37)

File: armsrc/lfsampling.c
Comment (iceman1001) [reply armsrc/lfsampling.c:492]: I guess its something about the trigger threshold behavior text that isn't clear enough?

Comment (wh201906) [reply armsrc/lfsampling.c:492]: I think it's better to leave this unchanged

File: armsrc/lfsampling.c
Comment (wh201906) [reply armsrc/lfsampling.c]: The `bits_per_sample` is from `config.bits_per_sample` which has been validated in `setSamplingConfig()`, so I think there is no need to check it again.

**Comment by wh201906** [comment]
I've cleaned up the commit history and made a new PR #2173.

**Pull request closed without merging**
